pkg/operator/workloads/logs.go

```diff
@@ -93,105 +141,140 @@ func StreamFromCloudWatch(podCheckCancel chan struct{}, appName string, podLabel
-					closeSocket(socket)
+				if !awslib.CheckErrCode(err, cloudwatchlogs.ErrCodeResourceNotFoundException) {
+					writeAndCloseSocket(socket, "error encountered while fetching logs from cloudwatch: "+err.Error())
 					continue
 				}
 			}
 
-			newEvents := strset.New()
+			lastLogTimestampMillis := TimeToMillis(lastLogTime)
 			for _, logEvent := range logEventsOutput.Events {
 				var log FluentdLog
 				json.Unmarshal([]byte(*logEvent.Message), &log)
-
-				if !previousEvents.Has(*logEvent.EventId) {
+				if !eventCache.Has(*logEvent.EventId) {
 					socket.WriteMessage(websocket.TextMessage, []byte(log.Log))
-					if *logEvent.Timestamp > lastTimestamp {
-						lastTimestamp = *logEvent.Timestamp
+					if *logEvent.Timestamp > lastLogTimestampMillis {
+						lastLogTimestampMillis = *logEvent.Timestamp
 					}
+					eventCache.Add(*logEvent.EventId)
 				}
-				newEvents.Add(*logEvent.EventId)
 			}
 
+			lastLogTime = MillisToTime(lastLogTimestampMillis)
 			if len(logEventsOutput.Events) == maxLogLinesPerRequest {
-				socket.WriteMessage(websocket.TextMessage, []byte("---- Showing at most "+s.Int(maxLogLinesPerRequest)+" lines. Visit AWS cloudwatch logs console and search for \""+prefix+"\" in log group \""+config.Cortex.LogGroup+"\" for complete logs ----"))
-				lastTimestamp = endTime
+				writeString(socket, "---- Showing at most "+s.Int(maxLogLinesPerRequest)+" lines. Visit AWS cloudwatch logs console and search for \""+prefix+"\" in log group \""+config.Cortex.LogGroup+"\" for complete logs ----")
+				lastLogTime = MillisToTime(endTime)
 			}
 
-			previousEvents = newEvents
-			timer.Reset(pollPeriod * time.Millisecond)
+			timer.Reset(pollPeriod)
+		}
+	}
+}
+
+func MillisToTime(epochMillis int64) time.Time {
+	seconds := epochMillis / 1000
+	millis := epochMillis % 1000
+	return time.Unix(seconds, millis*int64(time.Millisecond))
+}
+
+func TimeToMillis(t time.Time) int64 {
+	return t.UnixNano() / int64(time.Millisecond)
+}
+
+func getLogStreams(logGroupName string) (strset.Set, error) {
+	describeLogStreamsOnput, err := config.AWS.CloudWatchLogsClient.DescribeLogStreams(&cloudwatchlogs.DescribeLogStreamsInput{
+		Descending:   aws.Bool(true),
+		LogGroupName: aws.String(logGroupName),
+		OrderBy:      aws.String(cloudwatchlogs.OrderByLastEventTime),
+		Limit:        aws.Int64(50),
+	})
+	if err != nil {
+		if !awslib.CheckErrCode(err, cloudwatchlogs.ErrCodeResourceNotFoundException) {
+			return nil, err
 		}
+		return nil, nil
+	}
+
+	streams := strset.New()
+
+	for _, stream := range describeLogStreamsOnput.LogStreams {
+		streams.Add(*stream.LogStreamName)
 	}
+	return streams, nil
 }
 
-func getPrefix(searchLabels map[string]string) (string, error) {
+func getPodStartTime(searchLabels map[string]string) (time.Time, error) {
 	pods, err := config.Kubernetes.ListPodsByLabels(searchLabels)
 	if err != nil {
-		return "", err
+		return time.Now(), err
```

Comment:
Typically it's preferred to return the zero value when there's an error, so in this case `time.Time{}`